lib/ui/painting.dart

```diff
@@ -3768,13 +3768,15 @@ class FragmentProgram extends NativeFieldWrapperClass1 {
   ///
   /// [A current specification of valid SPIR-V is here.](https://github.com/flutter/engine/blob/master/lib/spirv/README.md)
   /// SPIR-V not meeting this specification will throw an exception.
-  ///
-  /// Performance of shader-compilation is platform dependent and is not
-  /// well-specified.  Because of this, it is reccommended to construct
-  /// `FragmentProgram` asynchronously, outside of a widget's `build`
-  /// method; this will minimize the chance of UI jank.
+  static Future<FragmentProgram> compile({
```

Comment:
I'll defer to @zanderso on whether this async construction is asynchronous enough. Minimally, declaring this static method async means it will return right away, but I don't have a good sense of when that additional work in the transpile method will happen and whether or not it can be broken up into chunks to prevent stalling future frames.

Reply:
Yeah, this won't do the trick. An `async` method runs synchronously until it hits the first `await`.

Something like this will actually run the computation on a subsequent turn of the event loop:

```dart
static Future<FragmentProgram> compile({
  required ByteBuffer spirv,
  bool debugPrint = false,
}) {
  return Future<FragmentProgram>(() => FragmentProgram._(spirv: spirv, debugPrint: debugPrint));
}
```

https://api.dart.dev/dev/2.15.0-285.0.dev/dart-async/Future/Future.html

It might need to be

https://api.dart.dev/dev/2.15.0-285.0.dev/dart-async/Future/Future.microtask.html

to avoid causing problems for frame workloads. @iskakaushik will know for sure.

Reply:
Is the event loop a good choice for this? We wouldn't want to stall events while this is happening.

Reply:
I'm not remembering all the work that happens during a compile, whether it's pure Dart code or whether parts of it are C++, but if it may cause frames to be delayed even when kicked off asynchronously, then the work will need to go on another thread. @iskakaushik's scheduling work will help avoid doing these compiles while frames are scheduled, but if that turns out not to be sufficient, then we'll have to e.g. get the Dart code for the compile running on a background isolate. I think the urgency of that optimization should be guided by benchmarks.

Reply:
My mental model when I made the suggestion was something along the lines of what happens when you try to load an image frame via ui.Codec.

Reply:
Is it sufficient for this PR to put the API in place? It appears that for `ui.Codec` the threading is done by C++, which would be fine here as well. We could add that once benchmarks indicate an issue if we add a compatible API here.

Reply:
I think we should move forward with this change as-is, and move the work to a thread if benchmarks and/or user experiences indicate there is a problem.

Reply:
@zanderso including the suggestion you gave above that kicks it until later, or literally as is?

Reply:
Sorry, I did not mean literally as-is. I meant including my suggestion above. Thanks, @flar.

Reply:
That sounds good to me too.